### PR TITLE
feat: Ingress replica CRD

### PR DESCRIPTION
**Pull request by zreigz**

### CODE REVIEW

File: api/v1alpha1/ingressreplica_types.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	// Spec of the IngressReplica
+	// +kubebuilder:validation:Required
+	Spec IngressReplicaSpec `json:"spec"`
+
+	// Status of the IngressReplica
+	// +kubebuilder:validation:Optional
+	Status Status `json:"status,omitempty"`
+}
+
+type IngressReplicaSpec struct {
```
Comment (michaeljguarino) [inline api/v1alpha1/ingressreplica_types.go:39]: We should also support overriding ingress class here

File: internal/controller/ingressreplica_controller.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		Spec: networkv1.IngressSpec{
+			IngressClassName: oldIngress.Spec.IngressClassName,
+			DefaultBackend:   oldIngress.Spec.DefaultBackend,
+		},
+	}
+	updateIngress(ingressReplica, newIngress, oldIngress)
+	return newIngress
+}
+
+func updateIngress(ingressReplica *v1alpha1.IngressReplica, newIngress *networkv1.Ingress, oldIngress *networkv1.Ingress) {
```
Comment (michaeljguarino) [inline internal/controller/ingressreplica_controller.go:127]: I think we should have some mechanism to merge in new annotations.  I'm on the fence as to whether we should simply take any annotation from the ingress replica itself and override the annotations on the old ingress or have an attribute on the spec.  Sort of leaning towards the first option.

File: api/v1alpha1/ingressreplica_types.go (hunk 1 above)
Comment (zreigz) [reply api/v1alpha1/ingressreplica_types.go:39]: done

File: internal/controller/ingressreplica_controller.go (hunk 2 above)
Comment (zreigz) [reply internal/controller/ingressreplica_controller.go:126]: I have used `lo.Assign` method which merges multiple maps from left to right.
```
newIngress.Annotations = lo.Assign(newIngress.Annotations, oldIngress.Annotations, ingressReplica.Annotations)
```

**Comment by zreigz** [comment]
@michaeljguarino PTAL

**Review by michaeljguarino — APPROVED** [review]

**Pull request merged**